### PR TITLE
txn: finish oracle

### PR DESCRIPTION
**Pull request by GanZiheng**
Based on
- https://github.com/tikv/agatedb/pull/67
- https://github.com/dgraph-io/badger/blob/45bca18f24ef5cc04701a1e17448ddfce9372da0/txn.go

Signed-off-by: GanZiheng [ganziheng98@gmail.com](mailto:ganziheng98@gmail.com)

### CODE REVIEW

File: src/ops/oracle.rs [hunk 1]
```diff
@@ -1,30 +1,312 @@
...
+                .any(|committed_txn| {
+                    reads
+                        .iter()
+                        .any(|read| committed_txn.conflict_keys.contains(read))
+                })
+        }
+    }
+}
+
 pub struct Oracle {
```
Comment (Connor1996) [inline src/ops/oracle.rs:72]: What's the oracle used for?

Comment (GanZiheng) [reply src/ops/oracle.rs:72]: It is used for transaction to get timestamp when reading or committing.

Comment (Connor1996) [reply src/ops/oracle.rs:72]: add a comment

File: src/levels/tests.rs [hunk 2]
```diff
@@ -31,6 +31,9 @@ where
     F: FnOnce(&mut LevelsController),
     F: Send + 'static,
 {
+    // So as to set and get discard timestamp by self.
+    opts.managed_txns = true;
```
Comment (Connor1996) [inline src/levels/tests.rs:35]: set it in `test_options()`?

File: src/ops/oracle.rs [hunk 3]
```diff
@@ -1,30 +1,312 @@
...
+            oracle.done_commit(commit_ts);
+        }
+
+        oracle.txn_mark.wait_for_mark(100);
+
+        oracle.stop();
+    }
+
+    #[test]
+    fn test_detect_confict() {
```
Comment (Connor1996) [inline src/ops/oracle.rs:266]: confict -> conflict

File: src/ops/oracle.rs [hunk 4]
```diff
@@ -1,30 +1,312 @@
...
+    discard_ts: u64,
+    /// Contains all committed writes (contains fingerprints of keys written
+    /// and their latest commit counter).
+    committed_txns: Vec<CommittedTxn>,
+    last_cleanup_ts: u64,
+}
+
+impl CommitInfo {
+    fn cleanup_committed_transactions(&mut self, is_managed: bool, read_mark: Arc<WaterMark>) {
+        let max_read_ts = {
```
Comment (Connor1996) [inline src/ops/oracle.rs:30]: unnecessary bracket

**Comment by skyzh** [comment]
I think we can decide whether to have the "managed mode" in agatedb. I believe systems built upon agatedb generally will have their own timestamp and way of watermark management. We can simply design agatedb as a user-timestamp system, and only have managed mode. This way we don't need conflict detector.

**Comment by GanZiheng** [comment]
> I think we can decide whether to have the "managed mode" in agatedb. I believe systems built upon agatedb generally will have their own timestamp and way of watermark management. We can simply design agatedb as a user-timestamp system, and only have non-managed mode. This way we don't need conflict detector.

If I understand correctly, systems with their own timestamp generator could just set timestamp in managed mode?

**Comment by Connor1996** [comment]
> I think we can decide whether to have the "managed mode" in agatedb. I believe systems built upon agatedb generally will have their own timestamp and way of watermark management. We can simply design agatedb as a user-timestamp system, and only have non-managed mode. This way we don't need conflict detector.

We can first port everything without tweaks, then truncate the unnecessary part.

**Review by zhangjinpeng1987 — APPROVED** [review]
LGTM

**Review by Connor1996 — APPROVED** [review]
LGTM

**Pull request merged**